components/daffy/daffy.app.mjs

```diff
@@ -0,0 +1,138 @@
+import { axios } from "@pipedream/platform";
+import { defineApp } from "@pipedream/types";
+
+export default defineApp({
+  type: "app",
+  app: "daffy",
+  propDefinitions: {
+    causeId: {
+      type: "string",
+      label: "Cause",
+      description: "The ID of the main cause.",
+      async options() {
+        const data = await this.listCauses();
+
+        return data.map(({
+          id: value, name: label,
+        }) => ({
+          label,
+          value,
+        }));
+      },
+    },
+    ein: {
+      type: "string",
+      label: "EIN",
+      description: "The Employer Identification Number of the non-profit.",
+      async options({
+        page, causeId,
+      }) {
+        const { items: data } = await this.listNonProfits({
+          params: {
+            cause_id: causeId,
+            page: page + 1,
+          },
+        });
+
+        return data.map(({
+          ein: value, name: label,
+        }) => ({
+          label,
+          value,
+        }));
+      },
+    },
+  },
+  methods: {
+    _apiUrl() {
+      return "https://public.daffy.org/v1";
+    },
+    _getHeaders() {
+      return {
+        "X-Api-Key": this.$auth.api_key,
+      };
+    },
+    async _makeRequest({
+      $ = this, path, ...opts
+    }) {
+      const config = {
+        url: `${this._apiUrl()}/${path}`,
+        headers: this._getHeaders(),
+        ...opts,
+      };
+
+      return axios($, config);
+    },
+    async createDonation(args = {}) {
+      const userId = await this.getMe();
+      return this._makeRequest({
+        method: "POST",
+        path: `users/${userId}/donations`,
+        ...args,
+      });
+    },
+    listNonProfits(args = {}) {
+      return this._makeRequest({
+        path: "non_profits",
+        ...args,
+      });
+    },
+    async listCauses(args = {}) {
+      const userId = await this.getMe();
+      return this._makeRequest({
+        path: `users/${userId}/causes`,
+        ...args,
+      });
+    },
+    async listDonations(args = {}) {
+      const userId = await this.getMe();
+      return this._makeRequest({
+        path: `users/${userId}/donations`,
+        ...args,
+      });
+    },
+    async getMe(args = {}) {
+      const { id: userId } = await this._makeRequest({
+        path: "users/me",
+        ...args,
+      });
+
+      return userId;
+    },
+    async *paginate({
+      fn, params = {}, maxResults = null,
+    }) {
+      let firstPage = false;
```

Comment:
This pagination goes from the last page to first, so it stops when it reaches the first page.

Reply:
Correct, but since `firstPage` starts out false, and it isn't updated the first time through:
```
if (firstRequest) {
   page = last;
   firstRequest = 0;
}
```
Won't it be false when it reaches `while (firstPage)` and exit?

Reply:
once this paginate starts from last page to first page, when it reaches the first page, `firstPage = !(currentPage === 1)` = false, it exits

Reply:
I may be missing something, but I'm not seeing how it gets to the point where it's paginating from the last page, if it exits after firstRequest. Just want to confirm, you were able to test this successfully?